protocol/src/protocol.semanticHighlighting.proposed.md

```diff
@@ -0,0 +1,134 @@
+#### Semantic Highlighting
+
+While the syntax highlighting is done on the client-side and can handle keywords, strings, and other low-level tokens from the grammar, it cannot adequately support complex coloring. Semantic highlighting information is calculated on the language server and pushed to the client as a notification. This notification carries information about the ranges that have to be colored. The desired coloring details are given as [TextMate scopes](https://manual.macromates.com/en/language_grammars) for each affected range. For the semantic highlighting information the following additions are proposed:
```

Comment:
It seems desirable to provide a way for the client to say "Send this range of text to a different language server if one is available", e.g. to provide a way to implement https://github.com/microsoft/vscode/issues/1751 (see https://github.com/microsoft/vscode/issues/77140#issuecomment-533953100 as well). This would allow, for example, an HTML language server to parse `onclick="event.preventDefault()"` and say "If a Javascript language server is available, hand it the `event.preventDefault()` text and let it parse that text instead". Other use cases where reparsing by a different language server would be useful include (but are not limited to):

* HTML in a string literal in a PHP file
* CSS in a `<style>` element in an HTML page
* SQL in a string literal in many different languages
* Markdown inside comments in many different languages
* Code embedded in Markdown in literate-programming documentation, Python `doctest` tests, etc

The common factor in all the above examples is that the "root" language server can identify the range of text in which a different language is embedded, and (usually) which language is embedded in that range. Therefore, one possible scope the language server may want to return is one with the meaning "Please make another semantic-highlighting request to a different language server that serves language XYZ, for this range of text".

To enable this scenario without making any change to the current protocol proposal, I propose a new root-level TextMate scope (in addition to `comment`, `keyword`, `string`, et al): `reparse`. A client that does not know about the `reparse` scope would ignore it; a client that understands it would understand a `reparse.language.markdown` scope as meaning "take the text in this range, and send it to a Markdown language server for parsing".

Since scopes (as seen below) are a 2D array, the scope that includes `reparse.language.markdown` could also include a second scope name that would be a fallback if no Markdown language server is available (e.g., if no editor extension is installed that provides Markdown parsing). This is most likely to be either a string or a comment. E.g., if entry 4 in the `scopes` array is `['reparse.language.markdown', 'string.quoted']` and entry 5 is `['reparse.language.markdown', 'comment.line.number-sign']`, then `"*bold*"` would return scope 4 while `# *bold*` would return scope 5.

There are other ways that https://github.com/microsoft/vscode/issues/1751 could be implemented in this API, such as adding a new `reparse` capability so that clients can signal whether or not they understand reparse messages, but that would involve a change in the API and might therefore hinder implementation by extensions that have already started to implement this. So I think a new top-level TextMate scope is probably the simplest way to achieve multi-language parsing.

Reply:
Highlighting and injection seems orthogonal to me. You'd want JS code completion in `"event.preventDefault()"`, and it would be strange if completion would have to block on syntax highlighting of the whole file. 

Rather, we need a special "document/languageInjections" request/notification, which returns something like this:

```TypeScript
interface InjectionInfo {
    language: string,
    range: Range,
    // We often want to strip leading indent or `//` from the injected fragment, 
    // `rangeMapping` should do that. An example would be like this
    // ```javascript
    // console.log({
    //   key: "<- that '//' is not a part of javascript snippet!"
    // })
    // ```
    rangeMapping?: Map<Range, Range>, 
    // Sometimes, just removing some ranges is not enough, 
    // a full string escaping might be required
    //
    //    onclick="console.log(&quotHello World!&quot)"
    textContent?: string
}
```

Reply:
>   // Sometimes, just removing some ranges is not enough, 
    // a full string escaping might be required
    //
    //    onclick="console.log(&quot;Hello World!&quot;)"
    textContent?: string

I don't think this is enough either - how does the `secondary` language server know how `textContent` was reached.

<details>
<summary>Demonstration</summary>
For the purposes of example take a pathological case where `&quot;` actually expands to `Hello World! ` in the `onclick` handler. A more realistic scenario would probably involve some backslashes

In this case, `console.log(&quot;Hello World!&quot;)` in the document would lead to a `textContent` of `console.log(Hello World! Hello World! Hello World! )`.

When hovering or typing at `Hello <|>World` in the document, how does the client know that the hover position know that the cursor position should be `31` [`console.log(Hello World! Hello <|>World! Hello World! )`] - from its perspective the one of the `&quot;`s may have expanded to each of ` ` or `Hello World! Hello World!`.
</details>

I think we need to make `rangeMapping` more general, i.e. replace it with a `TextEdit[]`, remove `textContent` and add an (optional?) `InjectionEncodeId`.

This would be passed in the `InjectionEncode` request, which is `InjectionEncodeParams -> TextEdit[]`, which turns 'in Injection' `TextEdit`s to 'out of Injection' `TextEdit`s.
```ts
interface InjectionEncodeParams {
	id: InjectionEncodeId, // Probably `number`
	edits: TextEdit[],
}
```

<details>
<summary>Worked Example</summary>

In the case of `onclick="console.log(&quot;Hello World!&quot;)"`, assuming that this is the entire document, the request would be:
`[server->client notification] "document/languageInjections"`

```jsonc
{
    "injections": [
        {
            "language": "javascript",
            "range": {
                "start": {
                    // Just after the open quote
                },
                "end": {
                    //Just after the close quote
                }
            },
            "edits": [
                {
                    "range": {
                        // Around the first `&quot;`
                    },
                    "newText": "\""
                },
                {
                    "range": {
                        // Around the second `&quot;`

                    },
                    "newText": "\""
                }
            ],
            "id": 0 // Some ID
        }
    ]
}
```

If the user then typed an edit (or accepted a completion) at "Hello <|>World", the client would make `InjectionEncode` request. This would (html) escape any quotes in the edit to be made for example.

N.B. I ran out of time to type out the request and responses in detail for this case.
</details>

Also, I think we would need some way of passing some sort of argument into this Injection, for example for Json Schemas for a sub file.